src/tools.c

```diff
@@ -206,218 +177,59 @@ static void cc_on_dialog_move_down_clicked(GtkButton *button, struct cc_dialog *
-
-		do
-		{
-			rv = g_io_channel_read_line(ioc, &msg, NULL, NULL, &err);
-			if (msg != NULL)
-			{
-				g_string_append(data->buffer, msg);
-				g_free(msg);
-			}
-			if (G_UNLIKELY(err != NULL))
-			{
-				geany_debug("%s: %s", G_STRFUNC, err->message);
-				g_error_free(err);
-				err = NULL;
-			}
-		} while (rv == G_IO_STATUS_NORMAL || rv == G_IO_STATUS_AGAIN);
-
-		if (G_UNLIKELY(rv != G_IO_STATUS_EOF))
-		{	/* Something went wrong? */
-			g_warning("%s: %s\n", G_STRFUNC, "Incomplete command output");
-		}
-	}
-	return FALSE;
-}
+	GError *error = NULL;
+	gchar *sel, *output, *errors;
+	gint status;
 
+	g_return_if_fail(doc != NULL && command != NULL);
+ 
+	if (! sci_has_selection(doc->editor->sci))
+		editor_select_lines(doc->editor, FALSE);
 
-static gboolean cc_iofunc_err(GIOChannel *ioc, GIOCondition cond, gpointer user_data)
-{
-	struct cc_data *data = user_data;
+	sel = sci_get_selection_contents(doc->editor->sci);
+	ui_set_statusbar(TRUE, _("Passing data and executing custom command: %s"), command);
 
-	if (cond & (G_IO_IN | G_IO_PRI))
+	if (spawn_with_capture(NULL, command, NULL, NULL, GTK_WINDOW(main_widgets.window), sel,
+		&output, &errors, &status, &error))
 	{
-		gchar *msg = NULL;
-		GString *str = g_string_sized_new(256);
-		GIOStatus rv;
-
-		do
+		if (*errors)
 		{
-			rv = g_io_channel_read_line(ioc, &msg, NULL, NULL, NULL);
-			if (msg != NULL)
-			{
-				g_string_append(str, msg);
-				g_free(msg);
-			}
-		} while (rv == G_IO_STATUS_NORMAL || rv == G_IO_STATUS_AGAIN);
-
-		if (!EMPTY(str->str))
-		{
-			g_warning("%s: %s\n", data->command, str->str);
+			g_warning("%s: %s\n", command, errors);
 			ui_set_statusbar(TRUE,
 				_("The executed custom command returned an error. "
 				"Your selection was not changed. Error message: %s"),
-				str->str);
-			data->error = TRUE;
-
+				errors);
+			g_free(errors);
 		}
-		g_string_free(str, TRUE);
-	}
-	data->finished = TRUE;
-	return FALSE;
-}
-
-
-static gboolean cc_replace_sel_cb(gpointer user_data)
-{
-	struct cc_data *data = user_data;
-
-	if (! data->finished)
-	{	/* keep this function in the main loop until cc_iofunc_err() has finished */
-		return TRUE;
-	}
-
-	if (! data->error && data->buffer != NULL && DOC_VALID(data->doc))
-	{	/* Command completed successfully */
-		sci_replace_sel(data->doc->editor->sci, data->buffer->str);
-	}
-
-	if (data->buffer)
-		g_string_free(data->buffer, TRUE);
-	g_slice_free1(sizeof *data, data);
-
-	return FALSE;
-}
-
-
-/* check whether the executed command failed and if so do nothing.
- * If it returned with a sucessful exit code, replace the selection. */
-static void cc_exit_cb(GPid child_pid, gint status, gpointer user_data)
-{
-	struct cc_data *data = user_data;
-
-	/* if there was already an error, skip further checks */
-	if (! data->error)
-	{
-#ifdef G_OS_UNIX
-		if (WIFEXITED(status))
+	#ifdef G_OS_WIN32
+		else if (status)
+	#else
+		else if (!WIFEXITED(status) || WEXITSTATUS(status) != EXIT_SUCCESS)
+	#endif
```

Comment:
> Maybe -1 if not WIFEXITED?

Yeah, it would probably be better indeed.

> Should be a function, status is evaluated twice.

yep

> But I'd like an okay on unifying the status handling in build, search and tools.

Sure.  Any reason why we would not want to?